File: common/src/main/java/generations/gg/generations/core/generationscore/world/item/PixelmonInteractions.java
package generations.gg.generations.core.generationscore.world.item;

//public class PixelmonInteractions {
//    private static final List<PixelmonInteraction> customInteractions = new ArrayList<>();
//    public static void registerCustomInteraction(PixelmonInteraction customInteraction) {
//        customInteractions.add(customInteraction);
//    }

//    public static void registerDefaultNonItemInteractions() {
//        registerCustomInteraction((pixelmonEntity, player, itemInHand) -> {
//            var data = pixelmonEntity.getPixelmonData();
//            var holder = PokeModRegistries.Pixelmon.SPECIES.getHolderOrThrow(data.getSpecies());

            /*if(itemInHand.getItem() == Items.SADDLE && PokeModRegistries.PIXELMON.registry().getTag(PokemonTags.SADDLEABLE).stream().anyMatch(a -> a.contains(holder))) {
                data.setSkin("saddle");

                return InteractionResult.CONSUME;
            } else {
                return InteractionResult.PASS;
            }*/
//            return InteractionResult.PASS;
//        });
//    }

//    public static InteractionResult triggerCustomInteraction(PixelmonEntity pixelmonEntity, Player player, ItemStack itemInHand) {
//        return customInteractions.stream()
//                .map(a -> a.interact(pixelmonEntity, player, itemInHand))
//                .filter(InteractionResult::consumesAction)
//                .findFirst()
//                .orElse(InteractionResult.PASS);
//    }
//
//    public interface PixelmonInteraction {
//        InteractionResult interact(PixelmonEntity pixelmonEntity, Player player, ItemStack itemInHand);
//
//        default boolean isConsumed() {
//            return true;
//        }
//    }
//}

File: common/src/main/kotlin/generations/gg/generations/core/generationscore/world/item/PixelmonInteractions.kt
package generations.gg.generations.core.generationscore.world.item

import com.cobblemon.mod.common.entity.pokemon.PokemonEntity
import dev.architectury.event.EventResult
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack

object PixelmonInteractions {
    private val customInteractions: MutableList<PixelmonInteraction> = ArrayList()
    fun registerCustomInteraction(customInteraction: PixelmonInteraction) {
        customInteractions.add(customInteraction)
    }

//    fun registerDefaultNonItemInteractions() {
//        registerCustomInteraction(PixelmonInteraction { pixelmonEntity: PokemonEntity, player: Player?, itemInHand: ItemStack? ->
//            val data: Unit = pixelmonEntity.getPixelmonData()
//            val holder: Unit = PokeModRegistries.Pixelmon.SPECIES.getHolderOrThrow(data.getSpecies())
//            InteractionResult.PASS
//        })
//    }

    fun triggerCustomInteraction(
        pixelmonEntity: PokemonEntity,
        player: Player,
        itemInHand: ItemStack
    ): EventResult {
        return customInteractions.stream()
            .map { a: PixelmonInteraction -> a.interact(pixelmonEntity, player, itemInHand) }
            .filter { obj: EventResult -> obj.interruptsFurtherEvaluation() }
            .findFirst()
            .orElse(EventResult.pass())
    }

    @JvmStatic
    fun process(entity: Entity, player: Player, stack: ItemStack): EventResult =
        if (entity is PokemonEntity && stack.item is PixelmonInteraction) (stack.item as PixelmonInteraction).interact(entity, player, stack)
        else EventResult.pass()

    interface PixelmonInteraction {
        fun interact(pokemon: PokemonEntity, player: Player, itemInHand: ItemStack): EventResult
        val isConsumed: Boolean
            get() = true
    }
}
